server/service/src/document/Programs.md

```diff
@@ -0,0 +1,144 @@
+
+```
+/p/patient1/Patient
+/p/patient1/HIVProgram
+/p/patient1/HIVEncounters/20220301T11:10
+/p/patient1/HIVEncounters/20220509T10:15
+```
+
+## Document json schema
+
+Each document uses a JSON schema which describes the document's data shape.
+The backend validates that an updated document fulfills the specified schema and otherwise rejects the document.
+
+## Special predefined document types
+
+Having a completely untyped data structures is not very useful and some common document formats are needed to provide some common functionality, e.g. listing all patients and encounters or doing a simple report in a generic ways.
+For this reason document types such as Patient, Programs or Encounter have a predefined base type, i.e. they define a common data shape for the respective type.
+However, clients are still able to extend these common document types.
+For example, an additional field can be added to a medical record.
+Note, that custom fields cannot be used in the generic part of the system but only in specialized views or reports.
+
+To make common documents easier to access they are "cached" in dedicated auxiliary DB tables with common data fields exposed as DB columns.
+Note, the underlying documents could still contain more custom data than stored in these aux tables.
+A dedicated DB table is only need if there is common program related functionality that couldn't be implement easily or efficiently using raw json documents.
+For example, a patient DB table could be used to efficiently search for common patient information such as name or date of birth or address, or an encounter DB table could be used to manage an schedule encounter events.
+
+## Derive custom JSON schema from base schema
+
+Common documents types have predefined commonly understood JSON schemas.
+To extend a base schema new fields can be added to the base schema in a backwards compatible way.
+Only adding new fields to a schema ensures that derived schemas still fullfil the base schema and thus can be handle generically.
+For example:
+
+```json
+{
+  // (base schema would need to be inlined though...)
+  "allOf": [{ "$ref": "/schema/base" }],
+  "properties": {
+    "baz": { "type": "string" }
+  }
+}
+```
+
+## Document registry
+
+In general documents are related to other documents, e.g. an encounter document is related to a certain type of program enrolment document.
+These relations of document types are stored in the document registry.
+The document registry also stores the json schema for a document type.
+For example, when editing a new document of a certain type the document registry dictates which json schema the document data must follow.
+
+Furthermore, the document registry contains a ui schema which defines how a document should be displayed.
+Currently there is exactly one ui schema for each document type but in the future it might be necessary to support multiple ui schema depending on the context in which a document is shown.
```

Comment:
```suggestion
If general documents are related to other documents, e.g. an encounter document is related to a certain type of program enrolment document, then these relations of document types are stored in the document registry.
The document registry also stores the json schema for a document type.
For example, when editing a new document of a certain type, the document registry dictates which json schema the document data must follow.

Furthermore, the document registry contains a UI schema which defines how a document should be displayed.
Currently there is exactly one UI schema for each document type but in the future, it might be necessary to support multiple UI schema depending on the context in which a document is shown.
```

Reply:
I wanted to say "In general" doesn't it work?

Reply:
You can but then it will be ... and these relations of

Reply:
Think it was ok in the original text, I pushed my changes...